src/AbstractJsonRepository.php

```diff
@@ -0,0 +1,569 @@
+<?php
+
+/*
+ * This file is part of the puli/repository package.
+ *
+ * (c) Bernhard Schussek <bschussek@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Puli\Repository;
+
+use Puli\Repository\Api\ChangeStream\ChangeStream;
+use Puli\Repository\Api\Resource\FilesystemResource;
+use Puli\Repository\Api\Resource\PuliResource;
+use Puli\Repository\Api\ResourceCollection;
+use Puli\Repository\Api\ResourceNotFoundException;
+use Puli\Repository\Api\UnsupportedResourceException;
+use Puli\Repository\Resource\Collection\ArrayResourceCollection;
+use Puli\Repository\Resource\DirectoryResource;
+use Puli\Repository\Resource\FileResource;
+use Puli\Repository\Resource\GenericResource;
+use Puli\Repository\Resource\LinkResource;
+use RuntimeException;
+use Webmozart\Assert\Assert;
+use Webmozart\Json\JsonDecoder;
+use Webmozart\Json\JsonEncoder;
+use Webmozart\PathUtil\Path;
+
+/**
+ * Abstract base for Path mapping repositories.
+ *
+ * @since  1.0
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+abstract class AbstractJsonRepository extends AbstractEditableRepository
+{
+    /**
+     * @var array
+     */
+    protected $json;
+
+    /**
+     * @var string
+     */
+    protected $baseDirectory;
+
+    /**
+     * @var string
+     */
+    private $path;
+
+    /**
+     * @var string
+     */
+    private $schemaPath;
+
+    /**
+     * @var JsonEncoder
+     */
+    private $encoder;
+
+    /**
+     * Creates a new repository.
+     *
+     * @param string            $path          The path to the JSON file. If
+     *                                         relative, it must be relative to
+     *                                         the base directory.
+     * @param string            $baseDirectory The base directory of the store.
+     *                                         Paths inside that directory are
+     *                                         stored as relative paths. Paths
+     *                                         outside that directory are stored
+     *                                         as absolute paths.
+     * @param ChangeStream|null $changeStream  If provided, the repository will
+     *                                         append resource changes to this
+     *                                         change stream.
+     * @param bool              $validateJson  Whether to validate the JSON file
+     *                                         against the schema. Slow but
+     *                                         spots problems.
```

Comment:
What do you think of enabling it by default on dev repository and disabling it by default on prod one.